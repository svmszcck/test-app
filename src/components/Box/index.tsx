import React from "react";
import { View } from "react-native";

const Box = () => {
  return <View></View>;
};

export default Box;
