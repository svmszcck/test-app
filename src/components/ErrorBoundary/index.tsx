import React, { Component, ErrorInfo } from "react";
import { View, StyleSheet } from "react-native";
import * as Updates from "expo-updates";
import { Text, Button } from "react-native-elements";

import { Theme } from "app_constants/colors";

class ErrorBoundary extends Component {
  state = {
    errorMsg: "",
  };

  static getDerivedStateFromError(error: Error) {
    return { errorMsg: error.toString() };
  }

  componentDidCatch(error: Error, info: ErrorInfo) {
    console.error("Uncaught error: ", error, info);
  }

  restartApp() {
    Updates.reloadAsync();
  }

  render() {
    if (this.state.errorMsg)
      return (
        <View style={styles.container}>
          <Text style={[styles.warning, { color: Theme.textBold }]}>
            Something went wrong :(
          </Text>
          <Button title="Restart App" onPress={this.restartApp} />
        </View>
      );
    return this.props.children;
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
  },
  warning: {
    fontSize: 20,
    marginBottom: 40,
  },
});

export default ErrorBoundary;
