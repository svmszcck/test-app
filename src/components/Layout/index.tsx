import React, { useMemo } from "react";
import {
  View,
  ActivityIndicator,
  StyleSheet,
  TouchableOpacity,
  useColorScheme,
} from "react-native";
import { Text } from "react-native-elements";
import { Ionicons } from "@expo/vector-icons";

import Colors from "constants/colors";

const Layout = ({
  children,
  isLoading = false,
  hasMenu = false,
  rightIcon,
  rightAction,
  navigation,
}: LayoutProps) => {
  const colorScheme = useColorScheme();
  const colors = useMemo(() => Colors[colorScheme], []);
  return (
    <View style={[styles.container, { backgroundColor: colors.primary }]}>
      {isLoading ? (
        <ActivityIndicator
          animating
          size="large"
          color={colors.text}
          style={styles.loading}
        />
      ) : (
        <>
          {hasMenu && (
            <View style={styles.header}>
              <TouchableOpacity
                style={{ width: 60, height: 60 }}
                onPress={() => navigation.goBack()}
              >
                <Ionicons size={30} color={colors.text} name="ios-arrow-back" />
              </TouchableOpacity>
              <TouchableOpacity style={styles.favorite} onPress={rightAction}>
                <Ionicons size={30} name={rightIcon} />
              </TouchableOpacity>
            </View>
          )}
          {children}
        </>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 20,
    paddingBottom: 0,
  },
  loading: {
    position: "absolute",
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  },
  header: {
    flexDirection: "row",
    marginBottom: 20,
  },
  favorite: {
    marginLeft: "auto",
    width: 30,
  },
});

type LayoutProps = {
  children: React.ReactNode;
  isLoading: boolean;
};

export default Layout;
