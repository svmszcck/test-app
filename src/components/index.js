export { default as Layout } from "./Layout";
export { default as Section } from "./Section";
export { default as List } from "./List";
export { default as Carousel } from "./Carousel";
export { default as ErrorBoundary } from "./ErrorBoundary";
