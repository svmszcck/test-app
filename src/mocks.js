import React from "react";
import { Alert } from "react-native";
import configureMockStore from "redux-mock-store";
import * as redux from "react-redux";
import { NavigationContainer } from "@react-navigation/native";
import { createStackNavigator } from "@react-navigation/stack";
import thunk from "redux-thunk";

const mockStore = configureMockStore([thunk]);
const Stack = createStackNavigator();
jest.spyOn(Alert, "alert");

export const state = {
  user: {
    name: "Onur",
    avatar: null,
    favorites: [
      { id: 13123, title: "Blabla", poster_path: "ewrwer", vote_average: 3.2 },
    ],
  },
  posts: {
    movie: {
      id: 12,
      title: "Rwer",
      poster_path: "ekjrhtkjert",
      vote_average: 5.6,
    },
    movies: [],
    genres: [
      {
        id: 1,
        name: "Action",
      },
      {
        id: 2,
        name: "Horror",
      },
      {
        id: 3,
        name: "Thriller",
      },
    ],
    popularMovies: [],
    searchedMovies: [],
    moviesByGenre: [
      { id: 12, title: "Hehehe", poster_path: "rtyrty", vote_average: 7.6 },
    ],
    genresLoading: false,
    moviesLoading: false,
    popularMoviesLoading: false,
    moviesByGenreLoading: false,
    movieLoading: false,
    isSearching: false,
  },
};

export const store = mockStore(() => state);

export const MockedNavigator = ({ component: Component, params = {} }) => {
  return (
    <NavigationContainer>
      <Stack.Navigator>
        <Stack.Screen name="MockedScreen" initialParams={params}>
          {(props) => <Component {...props} navigation={navigation} />}
        </Stack.Screen>
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export const navigation = {
  navigate: jest.fn(),
  goBack: jest.fn(),
};

export const mockUseSelector = jest.fn();
export const mockUseDispatch = jest.fn();

export const useDispatchSpy = jest.spyOn(redux, "useDispatch");
export const mockDispatchFn = jest.fn();
useDispatchSpy.mockReturnValue(mockDispatchFn);
