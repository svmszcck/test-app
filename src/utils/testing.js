import "react-native";
import Adapter from "enzyme-adapter-react-16";
import Enzyme from "enzyme";

const { JSDOM } = require("jsdom");

const url = "http://localhost";
const jsdom = new JSDOM("<!doctype html><html><body></body></html>", { url });
const { window } = jsdom;

jest.useFakeTimers();

export const setupEnv = () => {
  function copyProps(src, target) {
    Object.defineProperties(target, {
      ...Object.getOwnPropertyDescriptors(src),
      ...Object.getOwnPropertyDescriptors(target),
    });
  }

  global.window = window;
  global.document = window.document;
  global.navigator = {
    userAgent: "node.js",
  };
  copyProps(window, global);

  Enzyme.configure({ adapter: new Adapter() });
};
