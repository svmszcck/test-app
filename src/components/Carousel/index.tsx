import React from "react";
import {
  View,
  FlatList,
  StyleSheet,
  ActivityIndicator,
  GestureResponderEvent,
} from "react-native";
import { Text, Tile } from "react-native-elements";

import { IMAGE_URL } from "app_constants/api";
import Device from "app_constants/layout";
import { IMAGE_MEDIUM, POSTER_TEXT_LIMIT } from "app_constants/ui";
import Routes from "app_constants/routes";
import { trimText } from "utils/ui";
import { useColor } from "hooks";
import { Movie } from "types";

const posterWidth = Device.window.width / 2.5;
const posterHeight = posterWidth / 1.5;

const Carousel = ({
  elements = [],
  navigation,
  isLoading = false,
  loadMore,
}: CarouselType) => {
  const colors = useColor();
  return (
    <FlatList
      keyExtractor={(item) => item.id}
      data={elements}
      horizontal
      showsHorizontalScrollIndicator={false}
      renderItem={({
        item: { id, title, poster_path: posterPath, vote_average: voteAverage },
      }) => (
        <View style={styles.poster}>
          <Tile
            imageSrc={{
              uri: `${IMAGE_URL(IMAGE_MEDIUM)}/${posterPath}`,
            }}
            featured
            activeOpacity={0.8}
            height={posterHeight}
            width={posterWidth}
            onPress={() => navigation.navigate(Routes.MOVIE_DETAILS, { id })}
            imageContainerStyle={styles.posterImage}
            containerStyle={styles.posterContainer}
          />
          <Text style={[styles.posterTitle, { color: colors.text }]}>
            {title.length >= POSTER_TEXT_LIMIT
              ? trimText(title, POSTER_TEXT_LIMIT)
              : title}
          </Text>
        </View>
      )}
      ListFooterComponent={
        <View style={styles.posterLoadMore}>
          {isLoading ? (
            <ActivityIndicator animating size="large" color={colors.text} />
          ) : (
            <Text
              style={[styles.posterLoadMoreTxt, { color: colors.textBold }]}
              onPress={loadMore}
            >
              Load More
            </Text>
          )}
        </View>
      }
    />
  );
};

const styles = StyleSheet.create({
  poster: {
    flexDirection: "column",
    justifyContent: "center",
    alignItems: "center",
    marginRight: 20,
  },
  posterContainer: {
    justifyContent: "center",
    alignItems: "center",
    borderRadius: 10,
  },
  posterImage: {
    borderRadius: 10,
  },
  posterTitle: {
    marginTop: 5,
    fontSize: 15,
  },
  posterLoadMore: {
    justifyContent: "center",
    alignItems: "center",
    flex: 1,
    minWidth: 150,
    marginRight: 20,
  },
  posterLoadMoreTxt: {
    fontWeight: "bold",
    fontSize: 18,
  },
});

type CarouselType = {
  elements: Array<Movie>;
  navigation: any;
  isLoading: boolean;
  loadMore: (event: GestureResponderEvent) => void;
};

export default Carousel;
