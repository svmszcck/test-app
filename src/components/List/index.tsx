import React from "react";
import {
  View,
  ScrollView,
  ActivityIndicator,
  StyleSheet,
  FlatList,
} from "react-native";
import { Avatar, Button, Text, ListItem } from "react-native-elements";
import { isEmpty } from "lodash";
import { Ionicons } from "@expo/vector-icons";

import { IMAGE_URL } from "app_constants/api";
import Routes from "app_constants/routes";
import { IMAGE_SMALL } from "app_constants/ui";
import { DELAY_TIME, LOAD_MORE } from "app_constants/general";
import { delayTask } from "utils/ui";
import { Movie } from "types";
import { useColor } from "hooks";

import placeholder from "assets/images/movie.jpg";

const List = ({
  navigation,
  elements,
  hasAction = false,
  icon,
  rightAction,
  hasLoadMore = false,
  isLoading = false,
  loadMore,
  warning,
}: ListProps) => {
  const colors = useColor();

  return (
    <View>
      {!isLoading && isEmpty(elements) && warning && (
        <Text style={[styles.warning, { color: colors.textBold }]}>
          {warning}
        </Text>
      )}
      <ScrollView
        showsVerticalScrollIndicator={false}
        contentContainerStyle={styles.container}
      >
        <FlatList
          keyExtractor={(item) => item.id.toString()}
          data={elements}
          showsVerticalScrollIndicator={false}
          renderItem={({
            item: {
              id,
              title,
              poster_path: posterPath,
              vote_average: voteAverage,
            },
          }) => (
            <ListItem
              style={styles.item}
              onPress={() => navigation.navigate(Routes.MOVIE_DETAILS, { id })}
            >
              <Avatar
                size="large"
                source={
                  posterPath
                    ? {
                        uri: `${IMAGE_URL(IMAGE_SMALL)}/${posterPath}`,
                      }
                    : placeholder
                }
              />
              <ListItem.Content>
                <ListItem.Title>{title}</ListItem.Title>
                <ListItem.Subtitle style={{ color: colors.text }}>
                  Score: {voteAverage ? voteAverage : "-"}
                </ListItem.Subtitle>
              </ListItem.Content>
              {hasAction ? (
                <Ionicons
                  onPress={() => rightAction(id)}
                  size={30}
                  color={colors.text}
                  name={icon}
                />
              ) : (
                <ListItem.Chevron />
              )}
            </ListItem>
          )}
        />
        <View style={styles.bottomSection}>
          {hasLoadMore && !isLoading && !isEmpty(elements) && (
            <Button
              buttonStyle={[
                styles.loadMore,
                { backgroundColor: colors.secondary },
              ]}
              title={LOAD_MORE}
              onPress={() => delayTask(loadMore, DELAY_TIME)}
            />
          )}
          {isLoading && (
            <ActivityIndicator
              animating
              size="large"
              color={colors.text}
              style={styles.loading}
            />
          )}
        </View>
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    paddingBottom: 40,
  },
  warning: {
    marginTop: 30,
    marginLeft: "auto",
    marginRight: "auto",
    fontWeight: "bold",
  },
  bottomSection: {
    height: 80,
    alignItems: "center",
    justifyContent: "center",
  },
  loadMore: {
    margin: 20,
    backgroundColor: "white",
  },
  item: {
    marginBottom: 12,
  },
  loading: {
    marginTop: 20,
  },
});

type ListProps = {
  navigation: any;
  elements: Array<Movie>;
  hasAction: boolean;
  icon: string;
  rightAction: Function;
  hasLoadMore: boolean;
  isLoading: boolean;
  loadMore: Function;
  warning: string;
};

export default List;
