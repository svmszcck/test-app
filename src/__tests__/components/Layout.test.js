import React from "react";
import { ActivityIndicator, TouchableOpacity } from "react-native";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import { Layout } from "components";
import { store, navigation } from "mocks";
import { setupEnv } from "utils/testing";

setupEnv();

describe("Testing Layout Component", () => {
  const component = (
    <Provider store={store}>
      <Layout isLoading hasMenu />
    </Provider>
  );

  it("renders correctly", () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should show loading spinner", () => {
    const wrapper = mount(component);
    const loadingSpinner = wrapper.find(ActivityIndicator);

    expect(loadingSpinner.exists()).toBeTruthy();
  });

  it("should show the menu && go back", () => {
    const wrapper = mount(
      <Provider store={store}>
        <Layout isLoading={false} hasMenu navigation={navigation} />
      </Provider>
    );

    const backIcon = wrapper.find(TouchableOpacity).at(0);

    backIcon.props().onPress();

    expect(navigation.goBack).toHaveBeenCalled();
  });
});
