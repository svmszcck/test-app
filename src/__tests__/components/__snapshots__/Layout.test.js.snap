// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Layout Component renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "padding": 20,
        "paddingBottom": 0,
      },
      Object {
        "backgroundColor": "#E4F5F3",
      },
    ]
  }
>
  <ActivityIndicator
    animating={true}
    color="#5C7B8A"
    hidesWhenStopped={true}
    size="large"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</View>
`;

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "padding": 20,
        "paddingBottom": 0,
      },
      Object {
        "backgroundColor": "#E4F5F3",
      },
    ]
  }
/>
`;
